### DISCUSSION TITLE
Hosyond touch displays and tasmota

### DISCUSSION THREAD

**Discussion by 3xtant** [arendst/Tasmota#20759] [Q&A] [answered]
Hello, I am in a similar situation as other novices here, with a new display that has different combinations of display and touch controllers than those supported.

My display is a Hosyond 2.8 inch 240x320 IPS display with capacitive touch, which I bought from Amazon (US) https://www.amazon.com/dp/B0CMGF4Q68

The display is ILI9341, which works mostly great with universal display and the .ini file from the repo. But the touch driver is an FT6336G.

I have two issues, the first one is that my display shows inverted colors. I see that the .ini file has the :i option, but neither i1 nor i2 have any effect on the color display. I've resorted to enter inverted colors and use photoshop to create inverted color images to display, but that is a hacky and annoying way to work.

The second issue is the touch controller, which refuses to work. I am trying to use the FT6236 driver, but it actually causes the esp32 to reboot constantly and then revert back stock. The constant reboots have triggered WiFi issues that persist even after erasing the entire flash memory.

I am using an -lvgl-haspmota image to test the display, and a custom image based on tasmota32 for the display and touch. I do a full erase before switching between the two.

I believe there is a chance I am using GPIOs that are exposed by my ESP32-WROVER that shouldn't because they are used by the PSRAM, but I can't find the site where I read sometabout it. I am using the wiring hosyond provides in their examples, and their lvgl example does work for image and touch, so maybe the wiring issue is a tasmota issue, but I don't know when to look.

Any help is appreciated

**Comment by 3xtant** [comment]
@sfromis, as you mentioned in https://github.com/arendst/Tasmota/discussions/20754#discussioncomment-8510728

Is there a place when I can find the `display.ini` settings for the FT6236? I am hoping they work for my FT6336. They are the same family of controllers, except the 6336 has more capacity.

Also, can you tell me how to add the invert directive to my display.ini? The last line in the tasmota repo reads
`:i,20,21`
I need to set it to 21, but as I mentioned, issuing `DisplayText [i2]` does nothing. I tried doing it via the display.ini, but I added
`21,1,01`
This cause my display to stop responding and cause multiple reboots.

**Reply by sfromis** [reply]
I'm not into details with those touch controller chips, and can't do more than link to:
https://tasmota.github.io/docs/Displays/#universal-touch-driver

My knowledge is not much better around other parts of the `display.ini` file, but in the context of modifying the `:i,20,21` part, I don't even understand the logic or context of you adding `21,1,01`

The words around `:i` is that it is two display opcodes, which I think would have to be correct for the display controller type.

**Reply by 3xtant** [reply]
Okay, thanks for the reply. I have been re-reading the doc about the universal display/touch and I think I can try a few things to see if I can get touch to work.

About the `:i`, yes I agree my logic was flawed. It is not a flag, but an opcode. 🤦‍♂️ 
I spent some time reading the thread for the lvgl driver, they added the option to invert the colors because some displays need it (x21), there is a compiler flag to turn it on via config or in the header file, but there is a function that is supposed to allow the inversion at run time. It makes sense to me to add the line in the `display.ini` file, but I guess that particular opcode does nothing in my display, or is being overridden by something else.

**Reply by 3xtant** [reply]
For future users of hosyond displays: changing the order of the opcodes in the `:i` definition makes the display work correctly. That is, the line should read
`:i 21, 20`

**Comment by 3xtant** [comment]
I was unable to get my display to work with tasmota, which I realy wanted to do. Most of my devices are flashed with tasmota and this was the first place I looked when I wanted to create my own device.

Unfortunately, the universal touch control does not work with my FT6336 for reasons unknown to me. Just configuring the GPIO pins caused my device to reboot constantly. I created a tasmota driver for the touch controller, and even then, it refused to work.

After 4 days of struggle, I decided to try something different and checked esphome. It didn't look very user friendly, but it had drivers for both my display and touch controller. To my surprise, I had a working device about 30 minutes after reading the getting started section. Esphome lacks lvgl or haspmota-like stuff, so creating a UI is going to be harder, but everything worked almost instantly.

I looked briefly at the instructions for adding brand new devices to esphome and they seemed clear enough. Tasmota, in contrast, does not have easily accessible guidelines on how to port drivers or how to create new ones from scratch. Touch controllers have a very simple interface, fortunately for me, and I was able to look at other drivers to put my port together. I also saw in that process that the universal display driver has hard-coded stuff for  some controllers, which may explain why mine didn't work with it.

I am not saying esphome is better than tasmota, but I think their approach is much more modular and easier to assemble. Perhaps there is more work to do on the universal drivers for tasmota, and that could make this project a stronger contender for unknown hardware. I hope I can bring my new device back to tasmota, and I am not about reflashing all my stuff with esphome.

I care about this project. I can homestly say that without tasmota I would not have a smart home at all. I just want to give my sincere feedback because it can make this project the best it can be.

**Reply by gemu2015** [reply, marked as answer]
Universal Touch driver for FT6236
```
:UTI,FT6236,I1,38,-1,-1
RD A8
CP CD
RTF
RD A3
CP 36
RTF
RT
:UTT
RDM 00 16
MV 2 1
RT
:UTX
MV 3 2
RT
:UTY
MV 5 2
RT
#
```

**Reply by gemu2015** [reply]
the FT6X family of touch drivers are mostly compatible.
the only thing to adapt is vendor and chip ID which is checked on init.

check i2c on bus 1, at ic2 addr 38
:UTI,FT6236,I1,38,-1,-1
RD A8    read vendor ID at register A8
CP CD   compare with HEX CD
RTF        return when false
RD A3   read CHIP ID at register A3
CP 36   compare HEX 36
RTF       return when false
RT         return true

you may also simply skip checking these IDs
by just entering RT

**Reply by Beormund** [reply]
With the switch to uTouch my Wireless Tag WT32-SC01 display (with FT6336U touch chip) has stopped working with the latest release of Tasmota. I've tried adding the following section in my display.ini and omitted the vendor/chip id check as I do not know what they are. But still no joy:

```
:UTI,FT6336U,I1,38,-1,-1
RT
:UTT
RDM 00 16
MV 2 1
RT
:UTX
MV 3 2
RT
:UTY
MV 5 2
RT
#
```

**Reply by Beormund** [reply]
With the help of the FT6336U datasheet I finally got this working with the uTouch driver:
```
:UTI,FT6336U,I1,38,*,*
RD A0
CP 02
RTF
RT
:UTT
RDM 00 16
MV 2 1
RT
:UTX
MV 3 2
RT
:UTY
MV 5 2
RT
#
```
This should hopefully help anyone using WT32-SC01 / WT32-SC01 Plus displays.

**Reply by s-hadinger** [reply]
Thanks. I will add it to the documentation

**Reply by s-hadinger** [reply]
Would it make sense to add an autoconf file as well?

**Comment by Beormund** [comment]
> Would it make sense to add an autoconf file as well?

Here is the complete uDisplay/uTouch display.ini for WT32-SC01. What else do you need for the autoconf file? Thanks.

[display.ini.txt](https://github.com/arendst/Tasmota/files/15062312/display.ini.txt)

**Reply by sfromis** [reply]
Here's an example of an existing autoconf for a similar setup with a display:
https://github.com/tasmota/autoconf/tree/main/raw/esp32s3/SunTon-S3-Display-7inch

As you can see, in addition to `display.ini`, there is `init.bat` to set the template, and any other commands relevant to setting up the hardware to work. As templates vary by which ESP32 variant is being used, the autoconf files are categorized by this.

The `.autoconf` file is simply an uncompressed zip archive, renamed with the different extension, instead of `.zip`

**Reply by s-hadinger** [reply]
Let me at least add this file to the repo

**Comment by GasTurbineMan** [comment]
Not sure if this is the right place to put this info, but I have spent several hours trying to get a Hosyond 4" Capacitive Touch (480x320) display to work.  Two things:  1> Tasmota32-LVGL 15.3 seems to have an issue with Haspmota, I went back to 15.0.1 and Haspmota works again.  2> I never could get the FT6336 Touch to respond, finally after doing some research, I pulled the CTP_RST wire I had connected to ESP32 TS_RST pin and put in on 3.3v and touch works.  That is, on the DISPLAY, simply pull CTP_RST pin up to 3.3v, no need to connect it to the ESP32.

**Reply by findleybc** [reply]
@GasTurbineMan ,
Does Tasmota32-LVGL Latest (15.5 I think) work? I have spent several hours trying to get the touch on a Hosyond ESP32S3 2.8" development board with a FT6336G working. This is the touch section of my displaly.ini:

`:UTI,FT6336G,I1,38,-1,-1 
RD A0 
CP 02 
RTF 
RT 
:UTT 
RDM 00 16 
MV 2 1 
RT 
:UTX 
MV 3 2 
RT 
:UTY 
MV 5 2 
RT 
:M, FT6336G, 0, 240, 0, 320 
#`

**Reply by GasTurbineMan** [reply]
@findleybc , I can confirm Tasmota-LVGL 15.5 works fine on my displays. On my Hosyond 4" displays, the touch section is:

`:UTI,FT6336U,I1,38,*,*
RD A0
CP 02
RTF
RT
:UTT
RDM 00 16
MV 2 1
RT
:UTX
MV 3 2
RT
:UTY
MV 5 2
RT
#`

**Reply by findleybc** [reply]
Thanks for the help @GasTurbineMan ,
I did finally get it working. For my particular model (Hosyond ES3C28P) which has a FT6336**G** chip, the touch section is:

```
:UTI,FT6336G,I1,38,*,*
RD A0
CP 01
RTF
RT
:UTT
RDM 00 16
MV 2 1
RT
:UTX
MV 3 2
RT
:UTY
MV 5 2
RT
#
```

You module uses FT6336U and therefore the different CP value. 
The main reason it wasn't working was because I didn't use a template to define my GPIOs but instead set them in the module configuration screen. That worked for getting the display to work but in order to get touch working, the I2C, IRQ and RST for the chip must be in the template. I assume template is the preferred way to configure devices and therefore uTouch reads from there.

**Reply by sfromis** [reply]
I'd recommend wrapping those quoted blocks in lines with triple backtick, instead of single, to make formatting work better.
